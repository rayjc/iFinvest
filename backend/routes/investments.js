const express = require("express");

const ExpressError = require("../helpers/expressError");
const Investment = require("../models/investment");
const Portfolio = require("../models/portfolio");
const investmentSchema = require("../schemas/investment.json");
const investmentPatchSchema = require("../schemas/investmentPatch.json");
const { ensureLoggedIn } = require("../middleware/auth");
const validateJSON = require("../helpers/validateJson");
const ApiHelper = require("../helpers/apiHelper");
const calcInterest = require("../helpers/calcInterest");


const router = new express.Router();


router.get("/:id", ensureLoggedIn, async (req, res, next) => {
  try {
    const investment = await Investment.get(+req.params.id);
    return res.json({ investment });

  } catch (error) {
    return next(error);
  }
});


router.post("/", ensureLoggedIn, async (req, res, next) => {
  try {
    validateJSON(req.body, investmentSchema);

    const { initial_value, symbol, portfolio_id, start_date, end_date } = req.body;
    // check if portfolio belongs to user
    const portfolio = await Portfolio.get(portfolio_id);
    if (req.user.id !== portfolio.user_id) {
      throw new ExpressError(
        `Unauthorized; only user with id: ${portfolio.user_id} can modify ${portfolio.name}.`,
        401
      );
    }

    const investment = await Investment.create(
      initial_value, symbol, portfolio_id, start_date, end_date
    );

    return res.status(201).json({ investment });
  } catch (error) {
    return next(error);
  }
});


router.patch("/:id", ensureLoggedIn, async (req, res, next) => {
  try {
    validateJSON(req.body, investmentPatchSchema);

    const investment = await Investment.get(+req.params.id);
    // check if portfolio belongs to user
    const portfolio = await Portfolio.get(investment.portfolio_id);
    if (req.user.id !== portfolio.user_id) {
      throw new ExpressError(
        `Unauthorized; only user with id: ${portfolio.user_id} can modify ${portfolio.name}.`,
        401
      );
    }

    // update each field
    ["initial_value", "symbol", "start_date", "end_date"].forEach(function(field) {
      if (field in req.body) {
        investment[field] = req.body[field];
      }
    });

    const updatedInvestment = await investment.update();

    return res.json({ investment: updatedInvestment });

  } catch (error) {
    return next(error);
  }
});


router.delete("/:id", ensureLoggedIn, async (req, res, next) => {
  try {
    const investment = await Investment.get(+req.params.id);
    // check if portfolio belongs to user
    const portfolio = await Portfolio.get(investment.portfolio_id);
    if (req.user.id !== portfolio.user_id) {
      throw new ExpressError(
        `Unauthorized; only user with id: ${portfolio.user_id} can modify ${portfolio.name}`,
        401
      );
    }

    await investment.remove();

    return res.json({ message: `Investment(${investment.symbol}) deleted.` });
  } catch (error) {
    return next(error);
  }
});


router.get("/interest/:id", ensureLoggedIn, async (req, res, next) => {
  try {
    const investment = await Investment.get(+req.params.id);
    const { symbol, start_date, end_date } = investment;
    const initial = await ApiHelper.getSingleDataPoint(symbol, start_date);
    // get price at sell off if sold, otherwise get current price
    const final = end_date
      ? await ApiHelper.getSingleDataPoint(symbol, end_date)
      : await ApiHelper.getLastClose(symbol);

    // store initial price to db to avoid additional API calls
    investment.initial_price = initial.close;
    await investment.update();

    return res.json({
      symbol, start_date: initial.date, end_date: final.date,
      interest: calcInterest(final.close, initial.close)
    });

  } catch (error) {
    return next(error);
  }
});

module.exports = router;