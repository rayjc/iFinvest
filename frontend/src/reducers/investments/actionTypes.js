export const ADD_INVESTMENT_REQUEST = 'ADD_INVESTMENT_REQUEST';
export const ADD_INVESTMENT_FAILURE = 'ADD_INVESTMENT_FAILURE';
export const ADD_INVESTMENT_SUCCESS = 'ADD_INVESTMENT_SUCCESS';

export const UPDATE_INVESTMENT_REQUEST = 'UPDATE_INVESTMENT_REQUEST';
export const UPDATE_INVESTMENT_FAILURE = 'UPDATE_INVESTMENT_FAILURE';
export const UPDATE_INVESTMENT_SUCCESS = 'UPDATE_INVESTMENT_SUCCESS';

export const REMOVE_INVESTMENT_REQUEST = 'REMOVE_INVESTMENT_REQUEST';
export const REMOVE_INVESTMENT_FAILURE = 'REMOVE_INVESTMENT_FAILURE';
export const REMOVE_INVESTMENT_SUCCESS = 'REMOVE_INVESTMENT_SUCCESS';

export const LOAD_INVESTMENT_REQUEST = 'LOAD_INVESTMENT_REQUEST';
export const LOAD_INVESTMENT_FAILURE = 'LOAD_INVESTMENT_FAILURE';
export const LOAD_INVESTMENT_SUCCESS = 'LOAD_INVESTMENT_SUCCESS';