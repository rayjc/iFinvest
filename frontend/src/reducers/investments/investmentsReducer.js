import {
  ADD_INVESTMENT_FAILURE, ADD_INVESTMENT_REQUEST, ADD_INVESTMENT_SUCCESS,
  UPDATE_INVESTMENT_FAILURE, UPDATE_INVESTMENT_REQUEST, UPDATE_INVESTMENT_SUCCESS,
  REMOVE_INVESTMENT_FAILURE, REMOVE_INVESTMENT_REQUEST, REMOVE_INVESTMENT_SUCCESS,
  LOAD_INVESTMENT_FAILURE, LOAD_INVESTMENT_REQUEST, LOAD_INVESTMENT_SUCCESS,
} from './actionTypes';


const INIT_STATE = { investments: {}, error: null, isFetching: false };

function investmentsReducer(state = INIT_STATE, action) {
  switch (action.type) {
    case ADD_INVESTMENT_REQUEST:
    case UPDATE_INVESTMENT_REQUEST:
    case REMOVE_INVESTMENT_REQUEST:
    case LOAD_INVESTMENT_REQUEST:
      return { ...state, error: null, isFetching: true };

    case ADD_INVESTMENT_FAILURE:
    case UPDATE_INVESTMENT_FAILURE:
    case REMOVE_INVESTMENT_FAILURE:
    case LOAD_INVESTMENT_FAILURE:
      return { ...state, error: action.error, isFetching: false };

    case ADD_INVESTMENT_SUCCESS:
    case UPDATE_INVESTMENT_SUCCESS:
    case LOAD_INVESTMENT_SUCCESS:
      return {
        ...state,
        investments: {
          ...state.investments, [action.investment.id]: action.investment
        },
        isFetching: false
      };

    case REMOVE_INVESTMENT_SUCCESS:
      const { [action.id]: tobeRemoved, ...remainingInvestments } = state.investments;
      return {
        ...state, investments: { ...remainingInvestments }, isFetching: false
      };

    default:
      return state;
  }
}


export default investmentsReducer;