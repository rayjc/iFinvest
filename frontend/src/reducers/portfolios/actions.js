import {
  ADD_PORTFOLIO_FAILURE, ADD_PORTFOLIO_REQUEST, ADD_PORTFOLIO_SUCCESS,
  UPDATE_PORTFOLIO_FAILURE, UPDATE_PORTFOLIO_REQUEST, UPDATE_PORTFOLIO_SUCCESS,
  REMOVE_PORTFOLIO_FAILURE, REMOVE_PORTFOLIO_REQUEST, REMOVE_PORTFOLIO_SUCCESS,
  LOAD_PORTFOLIOS_FAILURE, LOAD_PORTFOLIOS_REQUEST, LOAD_PORTFOLIOS_SUCCESS,
  ADD_PORTFOLIO_INVESTMENT, REMOVE_PORTFOLIO_INVESTMENT,
} from './actionTypes';
import PortfolioApi from '../../api/portfolioApi';


function addPortfolioRequest() {
  return {
    type: ADD_PORTFOLIO_REQUEST,
  };
}

function addPortfolioFailure(error) {
  return {
    type: ADD_PORTFOLIO_FAILURE,
    error,
  };
}

function addPortfolioSuccess(portfolio) {
  return {
    type: ADD_PORTFOLIO_SUCCESS,
    portfolio,
  };
}

/**
 * Create a new portfolio on backend and store to redux.
 * @param {Number} name 
 */
function addPortfolio(name) {
  return async function(dispatch) {
    dispatch(addPortfolioRequest());

    try {
      const portfolio = await PortfolioApi.createPortfolio(name);
      dispatch(addPortfolioSuccess(portfolio));
    } catch (error) {
      console.error(error);
      dispatch(addPortfolioFailure(`Failed to create ${name} via backend API.`));
    }

  };
}


function updatePortfolioRequest() {
  return {
    type: UPDATE_PORTFOLIO_REQUEST,
  };
}

function updatePortfolioFailure(error) {
  return {
    type: UPDATE_PORTFOLIO_FAILURE,
    error,
  };
}

function updatePortfolioSuccess(id, name) {
  return {
    type: UPDATE_PORTFOLIO_SUCCESS,
    id,
    name,
  };
}

/**
 * Update the name of a portfolio of given id.
 * @param {Number} id
 * @param {String} name
 */
function updatePortfolio(id, name) {
  return async function(dispatch) {
    dispatch(updatePortfolioRequest());

    try {
      const portfolio = await PortfolioApi.updatePortfolio(id, name);
      dispatch(updatePortfolioSuccess(portfolio));
    } catch (error) {
      console.error(error);
      dispatch(updatePortfolioFailure(
        `Failed to update name of portfolio:${id} to ${name} via backend API.`
      ));
    }

  };
}


function removePortfolioRequest() {
  return {
    type: REMOVE_PORTFOLIO_REQUEST,
  };
}

function removePortfolioFailure(error) {
  return {
    type: REMOVE_PORTFOLIO_FAILURE,
    error,
  };
}

function removePortfolioSuccess(id) {
  return {
    type: REMOVE_PORTFOLIO_SUCCESS,
    id,
  };
}

/**
 * Remove a portfolio from store given id.
 * @param {Number} id
 */
function removePortfolio(id) {
  return async function(dispatch) {
    dispatch(removePortfolioRequest());

    try {
      await PortfolioApi.removePortfolio(id);
      dispatch(removePortfolioSuccess(id));
    } catch (error) {
      console.error(error);
      dispatch(removePortfolioFailure(
        `Failed to remove portfolio:${id} via backend API.`
      ));
    }

  };
}


function loadPortfoliosRequest() {
  return {
    type: LOAD_PORTFOLIOS_REQUEST,
  };
}

function loadPortfoliosFailure(error) {
  return {
    type: LOAD_PORTFOLIOS_FAILURE,
    error,
  };
}

function loadPortfoliosSuccess(portfolios) {
  return {
    type: LOAD_PORTFOLIOS_SUCCESS,
    portfolios,
  };
}

/**
 * Load portfolios onto store by fetching from backend API.
 */
function loadPortfolios() {
  return async function(dispatch) {
    dispatch(loadPortfoliosRequest());

    try {
      // fetch portfolios from backend
      const portfolios = await PortfolioApi.getPortfolios();
      dispatch(loadPortfoliosSuccess(portfolios));
    } catch (error) {
      console.error(error);
      dispatch(loadPortfoliosFailure("Failed to fetch portfolio from backend API."));
    }

  };
}


function addPortfolioInvestment(portfolioId, investmentId, symbol) {
  return {
    type: ADD_PORTFOLIO_INVESTMENT,
    portfolioId,
    investmentId,
    symbol,
  };
}

function removePortfolioInvestment(portfolioId, investmentId) {
  return {
    type: REMOVE_PORTFOLIO_INVESTMENT,
    portfolioId,
    investmentId,
  };
}


export {
  addPortfolio, removePortfolio, updatePortfolio, loadPortfolios,
  addPortfolioInvestment, removePortfolioInvestment
};