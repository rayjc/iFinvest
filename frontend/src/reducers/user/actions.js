import {
  LOAD_USER_FAILURE, LOAD_USER_REQUEST, LOAD_USER_SUCCESS,
  RESET_USER
} from './actionTypes';
import { decode } from "jsonwebtoken";
import UserApi from '../../api/userApi';


function loadUserRequest() {
  return { type: LOAD_USER_REQUEST };
}

function loadUserFailure(error) {
  return {
    type: LOAD_USER_FAILURE,
    error,
  };
}

function loadUserSuccess(user) {
  return {
    type: LOAD_USER_SUCCESS,
    user,
  };
}

function loadUser(token) {
  return async function(dispatch) {
    dispatch(loadUserRequest());

    try {
      const { id } = decode(token);
      const user = await UserApi.getUser(id);
      dispatch(loadUserSuccess(user));
    } catch (error) {
      console.error(error);
      dispatch(loadUserFailure("Failed to get user info from backend."));
    }

  };
}

function resetUser() {
  return { type: RESET_USER };
}


export { loadUser, resetUser };