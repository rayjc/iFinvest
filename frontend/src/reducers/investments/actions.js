import {
  ADD_INVESTMENT_FAILURE, ADD_INVESTMENT_REQUEST, ADD_INVESTMENT_SUCCESS,
  UPDATE_INVESTMENT_FAILURE, UPDATE_INVESTMENT_REQUEST, UPDATE_INVESTMENT_SUCCESS,
  REMOVE_INVESTMENT_FAILURE, REMOVE_INVESTMENT_REQUEST, REMOVE_INVESTMENT_SUCCESS,
  LOAD_INVESTMENT_FAILURE, LOAD_INVESTMENT_REQUEST, LOAD_INVESTMENT_SUCCESS,
} from './actionTypes';
import InvestmentApi from '../../api/investmentApi';
import { addPortfolioInvestment, removePortfolioInvestment } from '../portfolios/actions';


function addInvestmentRequest() {
  return { type: ADD_INVESTMENT_REQUEST };
}

function addInvestmentFailure(error) {
  return {
    type: ADD_INVESTMENT_FAILURE,
    error,
  };
}

function addInvestmentSuccess(investment) {
  return {
    type: ADD_INVESTMENT_SUCCESS,
    investment,
  };
}

/**
 * Create a new investment on backend and store to redux
 * @param {Object} newInvestment { initial_value, symbol, portfolio_id, 
 * start_date, end_date }
 */
function addInvestment(newInvestment) {
  return async function(dispatch) {
    dispatch(addInvestmentRequest());

    try {
      const investment = await InvestmentApi.createInvestment(newInvestment);
      const interest = await InvestmentApi.getInterest(investment.id);
      dispatch(addInvestmentSuccess({
        ...investment, ...interest, end_date: newInvestment.end_date
      }));
      // update investments in portfolios
      dispatch(addPortfolioInvestment(investment.portfolio_id, investment.id, investment.symbol));
    } catch (error) {
      console.error(error);
      dispatch(addInvestmentFailure("Failed to create a new investment via backend API."));
    }

  };
}


function updateInvestmentRequest() {
  return { type: UPDATE_INVESTMENT_REQUEST };
}

function updateInvestmentFailure(error) {
  return {
    type: UPDATE_INVESTMENT_FAILURE,
    error,
  };
}

function updateInvestmentSuccess(id, investment) {
  return {
    type: UPDATE_INVESTMENT_SUCCESS,
    id,
    investment,
  };
}

/**
 * Update an existing investment on backend and redux 
 * given an object with fields to be updated
 * @param {Number} id 
 * @param {Object} newInvestment 
 */
function updateInvestment(id, newInvestment) {
  return async function(dispatch) {
    dispatch(updateInvestmentRequest());

    try {
      const investment = await InvestmentApi.updateInvestment(id, newInvestment);
      const interest = await InvestmentApi.getInterest(investment.id);
      dispatch(updateInvestmentSuccess(
        id,
        { ...investment, ...interest, end_date: newInvestment.end_date }
      ));
    } catch (error) {
      console.error(error);
      dispatch(updateInvestmentFailure(`Failed to update investment:${id} via backend API.`));
    }

  };
}


function removeInvestmentRequest() {
  return { type: REMOVE_INVESTMENT_REQUEST };
}

function removeInvestmentFailure(error) {
  return {
    type: REMOVE_INVESTMENT_FAILURE,
    error,
  };
}

function removeInvestmentSuccess(id) {
  return {
    type: REMOVE_INVESTMENT_SUCCESS,
    id,
  };
}

/**
 * Remove an existing investment
 * @param {Number} id 
 * @param {Number} portfolioId 
 */
function removeInvestment(id, portfolioId) {
  return async function(dispatch) {
    dispatch(removeInvestmentRequest());

    try {
      dispatch(removePortfolioInvestment(portfolioId, id));
      await InvestmentApi.removeInvestment(id);
      dispatch(removeInvestmentSuccess(id));
    } catch (error) {
      console.error(error);
      dispatch(removeInvestmentFailure(`Failed to remove investment:${id} via backend API.`));
    }

  };
}


function loadInvestmentRequest() {
  return { type: LOAD_INVESTMENT_REQUEST };
}

function loadInvestmentFailure(error) {
  return {
    type: LOAD_INVESTMENT_FAILURE,
    error,
  };
}

function loadInvestmentSuccess(id, investment) {
  return {
    type: LOAD_INVESTMENT_SUCCESS,
    id,
    investment,
  };
}

/**
 * Fetch an investment from backend and store in redux
 * @param {Number} id 
 */
function loadInvestment(id) {
  return async function(dispatch) {
    dispatch(loadInvestmentRequest());

    try {
      const investment = await InvestmentApi.getInvestment(id);
      const interest = await InvestmentApi.getInterest(investment.id);
      dispatch(loadInvestmentSuccess(
        id,
        {
          ...investment, ...interest,
          end_date: investment.end_date ? investment.end_date.split('T')[0] : null
        }
      ));
    } catch (error) {
      console.error(error);
      dispatch(loadInvestmentFailure(`Failed to load investment:${id} via backend API.`));
    }

  };
}


export { addInvestment, updateInvestment, removeInvestment, loadInvestment };