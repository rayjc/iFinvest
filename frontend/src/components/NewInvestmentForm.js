import React, { useState } from 'react';
import { Button, Grid, InputAdornment, TextField } from '@material-ui/core';
import moment from 'moment';
import { useDispatch } from 'react-redux';
import { validateWeekday } from '../helpers/validate';
import { addInvestment } from '../reducers/investments/actions';
import SymbolField from './SymbolField';


const NewInvestmentForm = ({ portfolioId, handleClose }) => {
  const INIT_FORM_DATA = {
    "initial_value": "",
    "start_date": "",
    "end_date": "",
  };
  const dispatch = useDispatch();

  const [formData, setFormData] = useState(INIT_FORM_DATA);

  const [autoValue, setAutoValue] = useState({ symbol: "" });

  const handleChange = (evt) => {
    const { name, value } = evt.target;
    setFormData(f => ({ ...f, [name]: value }));
  };

  const handleSubmit = (evt) => {
    evt.preventDefault();
    const newInvestment = {
      ...formData,
      ...autoValue,
      initial_value: +formData.initial_value,
      portfolio_id: portfolioId
    };

    if (moment(formData.end_date).format('YYYY-MM-DD') === moment().format('YYYY-MM-DD')) {
      delete newInvestment.end_date;
    }
    dispatch(addInvestment(newInvestment));
    // dispatch(addPortfolioInvestment(portfolioId,));
    setFormData(INIT_FORM_DATA);
    handleClose();
  };

  return (
    <form onSubmit={handleSubmit}>
      <Grid container spacing={2} justify='center'>
        <Grid item xs={6}>
          <SymbolField value={autoValue} setValue={setAutoValue} />
        </Grid>
        <Grid item xs={6}>
          <TextField
            required
            id="initial-input"
            label="Initial Investment"
            name="initial_value"
            type="number"
            value={formData.initial_value}
            onChange={handleChange}
            InputProps={{
              startAdornment: <InputAdornment position="start">$</InputAdornment>,
              inputProps: { min: "0" },
            }}
          />
        </Grid>
        <Grid item xs={6}>
          <TextField
            required
            id="start-date"
            label="Start Date"
            type="date"
            name="start_date"
            value={formData.start_date}
            InputLabelProps={{
              shrink: true,
            }}
            onChange={handleChange}
            error={formData.start_date.length !== 0 && !validateWeekday(formData.start_date)}
            helperText={formData.start_date && !validateWeekday(formData.start_date) ? `Closed on weekend...` : ''}
            InputProps={{
              inputProps: {
                min: moment().subtract(5, "years").format("YYYY-MM-DD"),
                max: formData.end_date ? moment(formData.end_date).format('YYYY-MM-DD') : moment().format('YYYY-MM-DD'),
              },
            }}
          />
        </Grid>
        <Grid item xs={6}>
          <TextField
            required
            id="end-date"
            label="End Date"
            type="date"
            name="end_date"
            value={formData.end_date}
            InputLabelProps={{
              shrink: true,
            }}
            onChange={handleChange}
            // end_date could be null
            error={formData.end_date.length !== 0 && !validateWeekday(formData.end_date)}
            helperText={formData.end_date && !validateWeekday(formData.end_date) ? `Closed on weekend...` : ''}
            InputProps={{
              inputProps: {
                min: moment(formData.start_date).format("YYYY-MM-DD"),
                max: moment().format('YYYY-MM-DD')
              },
            }}
          />
        </Grid>
        <Grid item>
          <Button variant="contained" type="submit"
            disabled={!validateWeekday(formData.end_date) || !validateWeekday(formData.start_date)}>
            Add
          </Button>
        </Grid>
      </Grid>
    </form>
  );

};


export default NewInvestmentForm;