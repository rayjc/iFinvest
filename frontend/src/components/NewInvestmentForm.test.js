import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { fireEvent, render } from '@testing-library/react';
import NewInvestmentForm from './NewInvestmentForm';
import { useSelector, useDispatch } from 'react-redux';
import { addInvestment } from '../reducers/investments/actions';
import ApiHelper from '../api/apiHelper';


jest.mock('../api/apiHelper');

jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

jest.mock('../reducers/investments/actions', () => ({
  addInvestment: jest.fn(),
}));



beforeEach(function() {

  ApiHelper.mockClear();

  useSelector.mockImplementation((selectorFn) =>
    selectorFn({
      user: {
        user: {
          id: 0, username: "test", first_name: "test", last_name: "user",
          email: "test@test.com"
        },
        error: null, isFetching: false,
      },
      portfolios: {
        portfolios: [
          {
            id: 0, name: "alpha", user_id: 0, created_at: "2020-01-10",
            investments: [{ id: 0, symbol: "test" }]
          },
        ],
        error: null, isFetching: false,
      }
    })
  );
});


// smoke test
it("renders without crashing", function() {
  render(
    <MemoryRouter>
      <NewInvestmentForm portfolioId={0} handleClose={jest.fn()} />
    </MemoryRouter>
  );
});

// snapshot test
it("matches snapshot", function() {
  const { asFragment } = render(
    <MemoryRouter>
      <NewInvestmentForm portfolioId={0} handleClose={jest.fn()} />
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});

describe("form values can be update", function() {
  ApiHelper.request.mockResolvedValue({ stocks: [{ symbol: 'test' }] });

  it("symbol input can be changed", function() {
    const { getByLabelText } = render(
      <MemoryRouter>
        <NewInvestmentForm portfolioId={0} handleClose={jest.fn()} />
      </MemoryRouter>
    );

    const symbolInput = getByLabelText("Symbol", { exact: false });
    fireEvent.change(symbolInput, { target: { value: 'test' } });
    expect(symbolInput.value).toBe("test");
  });

  it("investment input can be changed", function() {
    const { getByLabelText } = render(
      <MemoryRouter>
        <NewInvestmentForm portfolioId={0} handleClose={jest.fn()} />
      </MemoryRouter>
    );

    const investmentInput = getByLabelText("Initial Investment", { exact: false });
    fireEvent.change(investmentInput, { target: { value: '100' } });
    expect(investmentInput.value).toBe("100");
  });

  it("startDate input can be changed", function() {
    const { getByLabelText } = render(
      <MemoryRouter>
        <NewInvestmentForm portfolioId={0} handleClose={jest.fn()} />
      </MemoryRouter>
    );

    const startDateInput = getByLabelText("Start Date", { exact: false });
    fireEvent.change(startDateInput, { target: { value: '2020-01-01' } });
    expect(startDateInput.value).toBe("2020-01-01");
  });

  it("endDate input can be changed", function() {
    const { getByLabelText } = render(
      <MemoryRouter>
        <NewInvestmentForm portfolioId={0} handleClose={jest.fn()} />
      </MemoryRouter>
    );

    const endDateInput = getByLabelText("End Date", { exact: false });
    fireEvent.change(endDateInput, { target: { value: '2020-01-01' } });
    expect(endDateInput.value).toBe("2020-01-01");
  });
});


it("Dispatches addInvestment() on submit", async function() {

  const mockedDispatch = jest.fn();
  useDispatch.mockReturnValue(mockedDispatch);

  const { getByLabelText, getByText, findByTestId } = render(
    <MemoryRouter>
      <NewInvestmentForm portfolioId={0} handleClose={jest.fn()} />
    </MemoryRouter>
  );

  const symbolInput = getByLabelText("Symbol", { exact: false });
  fireEvent.change(symbolInput, { target: { value: 'test' } });
  const investmentInput = getByLabelText("Initial Investment", { exact: false });
  fireEvent.change(investmentInput, { target: { value: '100' } });
  const startDateInput = getByLabelText("Start Date", { exact: false });
  fireEvent.change(startDateInput, { target: { value: '2020-07-14' } });
  const endDateInput = getByLabelText("End Date", { exact: false });
  fireEvent.change(endDateInput, { target: { value: '2020-07-15' } });

  const submitBtn = await findByTestId("add-investment-btn", { exact: false });
  fireEvent.click(submitBtn);

  expect(mockedDispatch).toHaveBeenCalled();
  expect(mockedDispatch).toHaveBeenCalledWith(addInvestment({
    symbol: "test",
    initial_value: "100",
    start_date: "2020-07-14",
    end_date: "2020-0715",
  }));
});