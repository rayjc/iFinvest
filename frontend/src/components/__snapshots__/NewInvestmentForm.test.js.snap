// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<form>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-justify-xs-center"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
    >
      <div
        aria-expanded="false"
        class="MuiAutocomplete-root MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon"
        name="symbol"
        role="combobox"
      >
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated Mui-required Mui-required"
            data-shrink="false"
            for="symbol-input"
            id="symbol-input-label"
          >
            Symbol
            <span
              aria-hidden="true"
              class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
            >
               *
            </span>
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
          >
            <input
              aria-autocomplete="list"
              aria-invalid="false"
              autocapitalize="none"
              autocomplete="off"
              class="MuiInputBase-input MuiInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd"
              id="symbol-input"
              required=""
              spellcheck="false"
              type="text"
              value=""
            />
            <div
              class="MuiAutocomplete-endAdornment"
            >
              <button
                aria-label="Clear"
                class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-clearIndicator MuiAutocomplete-clearIndicatorDirty"
                tabindex="-1"
                title="Clear"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-label="Open"
                class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-popupIndicator"
                tabindex="-1"
                title="Open"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
    >
      <div
        class="MuiFormControl-root MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink Mui-required Mui-required"
          data-shrink="true"
          for="initial-input"
          id="initial-input-label"
        >
          Initial Investment
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedStart"
        >
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionStart"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
            >
              $
            </p>
          </div>
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedStart"
            id="initial-input"
            min="0"
            name="initial_value"
            required=""
            type="number"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
    >
      <div
        class="MuiFormControl-root MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink Mui-required Mui-required"
          data-shrink="true"
          for="start-date"
          id="start-date-label"
        >
          Start Date
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input"
            id="start-date"
            max="2020-07-28"
            min="2015-07-28"
            name="start_date"
            required=""
            type="date"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
    >
      <div
        class="MuiFormControl-root MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink Mui-required Mui-required"
          data-shrink="true"
          for="end-date"
          id="end-date-label"
        >
          End Date
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input"
            id="end-date"
            max="2020-07-28"
            min="Invalid date"
            name="end_date"
            required=""
            type="date"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained"
        data-testid="add-investment-btn"
        tabindex="0"
        type="submit"
      >
        <span
          class="MuiButton-label"
        >
          Add
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</form>
`;
